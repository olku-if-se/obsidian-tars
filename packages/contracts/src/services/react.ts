/**
 * React Bridge service interface for managing React components in Obsidian
 */

import type { App } from 'obsidian'

// Generic React component type to avoid direct React dependency
export type ReactComponentType<T extends Record<string, unknown> = Record<string, unknown>> = (props: T) => unknown

export interface IReactBridge {
	/**
	 * Mount a React component into a container element
	 */
	mount<T extends Record<string, unknown>>(container: HTMLElement, component: ReactComponentType<T>, props: T): void

	/**
	 * Unmount a React component from a container element
	 */
	unmount(container: HTMLElement): void

	/**
	 * Unmount all mounted React components
	 */
	unmountAll(): void

	/**
	 * Check if a container has a mounted React component
	 */
	hasRoot(container: HTMLElement): boolean

	/**
	 * Get the number of currently mounted components
	 */
	getMountedCount(): number
}

export interface IReactBridgeManager {
	/**
	 * Get the React bridge instance
	 */
	getReactBridge(): IReactBridge

	/**
	 * Initialize the React bridge with Obsidian app
	 */
	initialize(app: App): void

	/**
	 * Clean up all React components
	 */
	dispose(): void

	/**
	 * Check if React bridge is ready
	 */
	isReady(): boolean
}
