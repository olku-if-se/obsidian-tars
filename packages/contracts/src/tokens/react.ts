export const ReactBridgeToken = Symbol('ReactBridge')
export const ReactBridgeManagerToken = Symbol('ReactBridgeManager')
export const StatusBarElementToken = Symbol('StatusBarElement')

export const AppToken = Symbol('App')
export const TarsPluginToken = Symbol('TarsPlugin')

// Type assertions to help TypeScript understand the token-service relationship
export type ReactBridgeToken = typeof ReactBridgeToken
export type ReactBridgeManagerToken = typeof ReactBridgeManagerToken
export type StatusBarElementToken = typeof StatusBarElementToken
export type AppToken = typeof AppToken
export type TarsPluginToken = typeof TarsPluginToken
