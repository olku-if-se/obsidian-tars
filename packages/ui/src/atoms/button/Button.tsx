import { forwardRef } from 'react'
import { useSemanticColors } from '../../components/atoms/ThemeProvider'
import styles from './Button.module.css'

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
	variant?: 'default' | 'danger' | 'primary'
	size?: 'sm' | 'md' | 'lg'
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
	({ variant = 'default', size = 'md', className, children, ...props }, ref) => {
		const colors = useSemanticColors()

		// Apply inline styles for theme colors as fallback
		const getVariantStyles = () => {
			switch (variant) {
				case 'primary':
					return {
						backgroundColor: colors.accent,
						color: colors.onAccent,
						borderColor: colors.accent
					}
				case 'danger':
					return {
						backgroundColor: colors.error,
						color: colors.onAccent,
						borderColor: colors.error
					}
				default:
					return {
						backgroundColor: colors.interactive,
						color: colors.text,
						borderColor: colors.border
					}
			}
		}

		const getSizeStyles = () => {
			switch (size) {
				case 'sm':
					return {
						padding: 'var(--size-2-1) var(--size-4-1)',
						fontSize: 'var(--font-ui-small)'
					}
				case 'lg':
					return {
						padding: 'var(--size-2-3) var(--size-4-4)',
						fontSize: 'var(--font-ui-large)'
					}
				default:
					return {
						padding: 'var(--size-2-2) var(--size-4-2)',
						fontSize: 'var(--font-ui-medium)'
					}
			}
		}

		const style = {
			...getVariantStyles(),
			...getSizeStyles(),
			borderRadius: 'var(--radius-m)',
			borderWidth: 'var(--border-width)',
			fontFamily: 'var(--font-interface)',
			fontWeight: 'var(--font-medium)',
			cursor: 'pointer',
			transition: 'background-color 0.2s ease, color 0.2s ease, border-color 0.2s ease'
		}

		return (
			<button
				ref={ref}
				className={`${styles.button} ${styles[variant]} ${styles[size]} ${className || ''}`}
				style={style}
				{...props}
			>
				{children}
			</button>
		)
	}
)

Button.displayName = 'Button'
