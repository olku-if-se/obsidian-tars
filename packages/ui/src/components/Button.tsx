import { forwardRef } from 'react'
import styles from './Button.module.css'

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
	variant?: 'default' | 'danger' | 'primary'
	size?: 'sm' | 'md' | 'lg'
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
	({ variant = 'default', size = 'md', className, children, ...props }, ref) => {
		return (
			<button ref={ref} className={`${styles.button} ${styles[variant]} ${styles[size]} ${className || ''}`} {...props}>
				{children}
			</button>
		)
	}
)

Button.displayName = 'Button'
