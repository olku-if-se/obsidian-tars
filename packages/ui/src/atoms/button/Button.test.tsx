import { render, screen, fireEvent } from '@testing-library/react'
import { describe, expect, it, vi, beforeEach } from 'vitest'
import { Button } from './Button'

describe('Button', () => {
	it('renders children correctly', () => {
		render(<Button>Click me</Button>)
		expect(screen.getByRole('button', { name: /click me/i })).toBeInTheDocument()
	})

	it('handles click events', () => {
		const handleClick = vi.fn()
		render(<Button onClick={handleClick}>Click me</Button>)

		fireEvent.click(screen.getByRole('button'))
		expect(handleClick).toHaveBeenCalledTimes(1)
	})

	it('applies variant classes', () => {
		render(<Button variant='primary'>Primary</Button>)
		const button = screen.getByRole('button')
		// CSS Modules generate scoped class names, so we check if the class contains 'primary'
		expect(button.className).toContain('primary')
	})

	it('applies default variant when none specified', () => {
		render(<Button>Default</Button>)
		const button = screen.getByRole('button')
		expect(button.className).toContain('button')
		expect(button.className).not.toContain('primary')
		expect(button.className).not.toContain('danger')
	})

	it('applies size classes', () => {
		render(<Button size='lg'>Large</Button>)
		const button = screen.getByRole('button')
		expect(button.className).toContain('lg')
	})

	it('applies custom className', () => {
		render(<Button className='custom-class'>Custom</Button>)
		const button = screen.getByRole('button')
		expect(button.className).toContain('custom-class')
	})

	it('is disabled when disabled prop is true', () => {
		render(<Button disabled>Disabled</Button>)
		expect(screen.getByRole('button')).toBeDisabled()
	})

	// New tests for React 19 compliance
	it('renders as different element when "as" prop is provided', () => {
		render(<Button as='a' href='https://example.com'>Link Button</Button>)
		const link = screen.getByRole('link')
		expect(link).toBeInTheDocument()
		expect(link).toHaveAttribute('href', 'https://example.com')
	})

	it('validates variant in development mode', () => {
		const originalEnv = process.env.NODE_ENV
		process.env.NODE_ENV = 'development'
		const consoleSpy = vi.spyOn(console, 'warn').mockImplementation(() => {})

		// @ts-ignore - intentionally testing invalid prop
		render(<Button variant='invalid'>Invalid</Button>)
		expect(consoleSpy).toHaveBeenCalledWith(
			expect.stringContaining('Invalid variant "invalid"')
		)

		consoleSpy.mockRestore()
		process.env.NODE_ENV = originalEnv
	})

	it('validates size in development mode', () => {
		const originalEnv = process.env.NODE_ENV
		process.env.NODE_ENV = 'development'
		const consoleSpy = vi.spyOn(console, 'warn').mockImplementation(() => {})

		// @ts-ignore - intentionally testing invalid prop
		render(<Button size='invalid'>Invalid</Button>)
		expect(consoleSpy).toHaveBeenCalledWith(
			expect.stringContaining('Invalid size "invalid"')
		)

		consoleSpy.mockRestore()
		process.env.NODE_ENV = originalEnv
	})

	// Test memoization behavior
	it('re-renders only when props change', () => {
		const handleClick = vi.fn()
		const { rerender } = render(
			<Button onClick={handleClick} variant='primary'>
				Test Button
			</Button>
		)

		// Rerender with same props
		rerender(
			<Button onClick={handleClick} variant='primary'>
				Test Button
			</Button>
		)

		// Button should still work correctly
		fireEvent.click(screen.getByRole('button'))
		expect(handleClick).toHaveBeenCalledTimes(1)
	})

	// Accessibility tests
	it('has proper accessibility attributes', () => {
		render(
			<Button disabled aria-label='Custom label'>
				Button Content
			</Button>
		)

		const button = screen.getByRole('button')
		expect(button).toHaveAttribute('aria-label', 'Custom label')
		expect(button).toBeDisabled()
	})

	// Snapshot test
	it('matches snapshot for different variants', () => {
		const { asFragment } = render(
			<div>
				<Button variant='default'>Default</Button>
				<Button variant='primary'>Primary</Button>
				<Button variant='danger'>Danger</Button>
				<Button size='sm'>Small</Button>
				<Button size='lg'>Large</Button>
			</div>
		)
		expect(asFragment()).toMatchSnapshot()
	})
})
