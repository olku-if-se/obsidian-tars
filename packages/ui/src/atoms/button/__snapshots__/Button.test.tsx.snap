// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > matches snapshot for different variants 1`] = `
<DocumentFragment>
  <div>
    <button
      class="_button_a9d9e1 _default_a9d9e1 _md_a9d9e1"
    >
      Default
    </button>
    <button
      class="_button_a9d9e1 _primary_a9d9e1 _md_a9d9e1"
    >
      Primary
    </button>
    <button
      class="_button_a9d9e1 _danger_a9d9e1 _md_a9d9e1"
    >
      Danger
    </button>
    <button
      class="_button_a9d9e1 _default_a9d9e1 _sm_a9d9e1"
    >
      Small
    </button>
    <button
      class="_button_a9d9e1 _default_a9d9e1 _lg_a9d9e1"
    >
      Large
    </button>
  </div>
</DocumentFragment>
`;
